### PR TITLE
Remove Sugar.js

### PR DESCRIPTION
**Pull request by Slayer95**
Remove Sugar.js

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ const path = require('path');
 // aren't
 
 try {
-	require('sugar');
+	require.resolve('sugar-deprecated');
```
Comment (Zarel) [inline app.js]: Wow, you made a package just for us. Nice.

**Comment by Zarel** [comment]
I'd prefer to completely remove sugar.js before merging this.

File: users.js [hunk 2]
```diff
@@ -1449,7 +1455,10 @@ class User {
 			};
 		}
 		this.send('|updatechallenges|' + JSON.stringify({
-			challengesFrom: Object.map(this.challengesFrom, 'format'),
+			challengesFrom: Object.keys(this.challengesFrom).reduce((prev, next) => {
+				prev[next] = this.challengesFrom[next].format;
+				return prev;
+			}, {}),
```
Comment (Zarel) [inline users.js]: I'm not sure I like the idea of fancy reduce usage. Would rather have a for-in loop, I think.

**Comment by Slayer95** [comment]
There is no more ``sugar`` in the public code base to the best of my knowledge.
There are things to do in the infraestructure and config though.

**Comment by Zarel** [comment]
Infrastructure? Config?

File: users.js [hunk 3]
```diff
@@ -270,7 +270,13 @@ function cacheGroupData() {
...
 			let inheritGroup = groups[groupData['inherit']];
 			if (cacheGroup(groupData['inherit'], inheritGroup)) {
-				Object.merge(groupData, inheritGroup, false, false);
+				// Add lower group permissions to higher ranked groups,
+				// preserving permissions specifically declared for the higher group.
+				Object.keys(inheritGroup || {}).reduce((groupData, key) => {
+					if (key in groupData) return groupData;
+					groupData[key] = inheritGroup[key];
+					return groupData;
+				}, groupData);
```
Comment (Zarel) [inline users.js]: Ditto this reduce.

File: tools.js [hunk 4]
```diff
@@ -1018,7 +1053,13 @@ module.exports = (() => {
...
 						// instead of overwriting entirely
 						delete data[dataType][key].inherit;
-						Object.merge(data[dataType][key], parentTypedData[key], false, false);
+
+						// Merge parent into children entry, preserving existing childs' properties.
+						Object.keys(parentTypedData[key] || {}).reduce((childEntry, subKey) => {
+							if (subKey in childEntry) return childEntry;
+							childEntry[subKey] = parentTypedData[key][subKey];
+							return childEntry;
+						}, data[dataType][key]);
```
Comment (Zarel) [inline tools.js]: And this reduce.

**Comment by Zarel** [comment]
reduce is nice for certain things like straight-up summation, but for-in is more performant and usually more readable...

**Comment by Slayer95** [comment]
Infraestructure concerns resolved with the introduction of ``Tools#toTimeStamp``. No config changes were needed after all. And reduce is RIP.

**Comment by Zarel** [comment]
Can you remove the rest of us from your crash e-mail list? :/

File: app.js (hunk 1 above)
Comment (Zarel) [inline app.js]: Although we might want to change this check to a check for SockJS?

File: package.json [hunk 5]
```diff
@@ -5,7 +5,8 @@
   "version": "0.10.2",
   "dependencies": {
     "sockjs": "0.3.15",
-    "sugar": "1.4.1"
+    "sugar-deprecated": "0.0.1",
+    "object.values": "1.0.3"
```
Comment (Zarel) [inline package.json]: We're not really decreasing our dependency count with this... can we just manually shim it in `tools.js`?

Comment (Slayer95) [inline package.json]: I wouldn't say ``count`` is the right metric for that.

**Comment by Slayer95** [comment]
I guess I can directly write to some location in ``logs`` and check it later?

**Comment by Zarel** [comment]
I mean, you're already testing it in your server. I want your server to stop emailing us. Main can email us all it likes.

File: package.json (hunk 5 above)
Comment (Zarel) [inline package.json]: :S

**Comment by Slayer95** [comment]
Ohhhhhhhhhhhhhhh! Apologies, that was while checking the config...

File: app.js (hunk 1 above)
Comment (Slayer95) [inline app.js]: That wouldn't automatically install new dependencies. I have always seen these instructions as intended to making sure that all the dependencies are installed.

**Pull request merged**